### ISSUE TITLE
Assess if a specific content exist in an array

### ISSUE THREAD

**Issue by Amachua** [stoplightio/spectral#624]
**User story.**
As a API designer, When I define a list of values in a specification, then I want to ensure that it does contains specific values.

**Is your feature request related to a problem?**
I've two main concerns with this request:
- When I define a new endpoint, I want to ensure that the authorization header is always defined in the header;
- Another technical constraint with an additional layer used in our CI/CD process.

**Describe the solution you'd like**
For the description, let's consider the following specification:
```YAML
openapi: 3.0.0

x-format-version: "1.0"

info:
  title: Dummy title
  description: Dummy description
  version: 1.0.0

paths:
  /resources:
    get:
      description: Dummy description
      responses:
        "401":
          description: All is good
      parameters:
        - name: authorization
          description: Bearer token. The token is formatted as a JWT. Format "Bearer JWT".
          in: header
          type: string
          required: true
      
    patch:
      description: Dummy description
      responses:
        "400":
          description: Something is missing here
```
For the GET request, everything is alright but for the patch, the authorization header isn't define --> an error must be prompted here.

I've imagine the following rule for that:
```
  authorization-header:
    type: validation
    severity: error
    recommended: true
    message: "Endpoint should have a 'authorization' header. Error: {{error}}"
    given: "$.paths.*[?( @property === 'get' || @property === 'put' || @property === 'post' || @property === 'delete' || @property === 'options' || @property === 'head' || @property === 'patch'  )].parameters*[?( @in === 'header' )]"
    then:
      function: propertyContentFound
      functionOptions:
        field: name
        match: ^authorization$
```

Also, I though that such feature could be used for the revamp of the `operation-2xx-response` rule. 

```
  operation-2xx-response-v2:
    type: validation
    severity: error
    recommended: true
    message: "Endpoint should have a 'authorization' header. Error: {{error}}"
    given: "$.paths.*[?( @property === 'get' || @property === 'put' || @property === 'post' || @property === 'delete' || @property === 'options' || @property === 'head' || @property === 'patch'  )].responses"
    then:
      function: propertyContentFound
      functionOptions:
        field: @key
        match: ^2[0-9]{2,2}$
```

If you need some more information on this point, I would be glad to help you with that. :)

**Comment by kylesykes** [comment]
We just solved a similar problem using Spectral v4, so hopefully this helps lead you in the right direction.  We ended up using a custom function to do what we wanted, although we have plans to contribute a more general standard function that achieves the same and I'm sure there's a cleaner JSON Path, but we are dealing with a lot of $ref's all over the place, especially when it comes to parameters.

Basically the JSONPath below gives us an array of objects (the parameters), and the custom function goes through each parameter and creates a new array of each parameter.name, then the custom function checks to see if (for that endpoint), it finds the desired name value (authorization) in that array.

Here's an altered (untested) version of what we're using.  I hope it helps you going forward, you'll likely have to tweak the JSONPath to meet your specific needs though.  This `exists` type of custom function has come in pretty handy for us, and we'll likely try to contribute it back as a standard function once we've covered most of the edge cases.  I'm sure it can be improved.  We have a more general `exists` function that can take in the thing you're looking for as a `functionOption`, but it's difficult to test for our particular case so we're satisfied with something like this for now.  

The trickiest part we found with this problem was we wanted Spectral to iterate over the particular `parameters` collection while simultaneously checking some parameter.name, but any attempts to try and get clever with JSONPath and return an array of arrays of strings (one array for each path/verb, and strings being the parameter.name) failed, so we resorted to just iterating over the parameter objects within a JS function.  I definitely think there's a use case for a more general function here, but we were unable to get any of the built in functions to handle it.

```yaml
requests-have-authorization-header:
    description: All endpoints have an authorization parameter in requests
    given: $.paths.*.get
    severity: error
    recommended: true
    then:
      field: parameters
      function: authorization_exists
```

And here's an (again, untested) custom function, `authorization_exists.js`
```js
module.exports = (targetVal) => {
    const valueExists = 'authorization';

    var references = [];

    for (var v of targetVal) {
      references.push(v['name']);
    }

    if (!references.includes(valueExists)) {
        return [
            {
              message: `The value {valueExists} was not found.`,
            },
          ];
    }

    return [];
  };
```

**Comment by Amachua** [comment]
Thank you for this answer @kylesykes, I do hope that we'll be able to do it one day directly with a JSONPath expression but for the time being, I'll discuss on this point directly with my technical lead (the input you gave us with the java-script will reduce our effort --> Thank you for this! :)).

**Comment by P0lip** [comment]
@Amachua
Can I consider this ticket closed? :)
If you have any questions, do let us know, please. Happy to answer any questions you might have.

**Comment by Amachua** [comment]
@P0lip, Yes you can consider this latter as closed! Also, I've found a workaround on another topic that might interest you (see below)

```YAML
  missing-authorization-header:
    type: style
    severity: error
    recommended: true
    formats: ["oas3"]
    description: Any authenticated operation must define the authorization header
    message: "Authorization header must be defined"
    given: "$.paths.*[?((@property === 'get' || @property === 'post' || @property === 'delete' || @property === 'options' || @property === 'patch' || @property === 'put') && (@.security))]"
    then:
      field: parameters
      function: schema
      functionOptions:
        schema:
          type: array
          contains:
            type: object
            properties:
              name:
                type: string
                pattern: ^authorization$
              in:
                type: string
                pattern: ^header$
```
